From: David Bauer <mail@david-bauer.net>
Date: Mon, 29 Jun 2020 12:24:01 +0200
Subject: ramips: add support for RAVPower RP-WD009

The RAVPower RP-WD009 is a batter-powered pocket sized router with SD
card lot and USB port.

Hardware
--------
CPU:   MediaTek MT7628AN
RAM:   64M DDR2
FLASH: 16M GigaDevices SPI-NOR
WLAN:  MediaTek MT7628AN 2T2R b/g/n
       MediaTek MT7610E  1T1R n/ac
ETH:   1x FastEthernet
SD:    SD Card slot
USB:   USB 2.0

Custom PMIC on the I2C bus (address 0x0a).

Installation
------------

1. Press and hold down the reset button.

2. Power up the Device. Keep pressing the reset button for 10
   more seconds until the Globe LED lights up.

3. Attach your Computer to the Ethernet port. Assign yourself the
   address 10.10.10.1/24.

4. Access the recovery page at 10.10.10.128 and upload the OpenWrt
   factory image.

5. The flashing will take around 1 minute. The device will reboot
   automatically into OpenWrt.

Signed-off-by: David Bauer <mail@david-bauer.net>
(cherry picked from commit e959048c128e4350b78df653be7cd7277787518e)
(cherry picked from commit 6481f21ffe2475c1b7d7ce4cdc428eded7bcf0ae)

diff --git a/target/linux/ramips/base-files/etc/board.d/02_network b/target/linux/ramips/base-files/etc/board.d/02_network
index f743ce851ad335ad1d81638ce5d2dda4e7df1620..d55965c9a39563f5318655ed519e493a4dea5c5c 100755
--- a/target/linux/ramips/base-files/etc/board.d/02_network
+++ b/target/linux/ramips/base-files/etc/board.d/02_network
@@ -159,6 +159,7 @@ ramips_setup_interfaces()
 	na930|\
 	pbr-d1|\
 	ravpower,wd03|\
+	ravpower,rp-wd009|\
 	tama,w06|\
 	tplink,tl-mr3020-v3|\
 	tplink,tl-wr802n-v4|\
diff --git a/target/linux/ramips/dts/RP-WD009.dts b/target/linux/ramips/dts/RP-WD009.dts
new file mode 100644
index 0000000000000000000000000000000000000000..d3db97fdeb0bc035024aed9f4f10910cc14f966c
--- /dev/null
+++ b/target/linux/ramips/dts/RP-WD009.dts
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include "mt7628an.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "ravpower,rp-wd009", "mediatek,mt7628an-soc";
+	model = "RAVPower RP-WD009";
+
+	aliases {
+		led-boot = &led_globe;
+		led-failsafe = &led_globe;
+		led-running = &led_globe;
+		led-upgrade = &led_globe;
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x4000000>;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_globe: globe {
+			label = "rp-wd009:white:globe";
+			gpios = <&gpio1 11 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan2 {
+			label = "rp-wd009:white:wlan2";
+			gpios = <&gpio1 12 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "phy0tpt";
+		};
+
+		wlan5 {
+			label = "rp-wd009:white:wlan5";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy1tpt";
+		};
+
+		sd_white {
+			label = "rp-wd009:white:sd";
+			gpios = <&gpio1 9 GPIO_ACTIVE_LOW>;
+		};
+
+		sd_red {
+			label = "rp-wd009:red:sd";
+			gpios = <&gpio0 3 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio1 14 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		/* Power interrupt on Pin 39 */
+
+		rfkill {
+			label = "rfkill";
+			gpios = <&gpio0 21 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RFKILL>;
+		};
+
+		backup {
+			label = "backup";
+			gpios = <&gpio1 8 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_COPY>;
+		};
+	};
+};
+
+&gpio0 {
+	mt7610-power {
+		gpio-hog;
+		gpios = <20 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "mt7610-power";
+	};
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			groups = "uart1", "wled_an", "p0led_an", "p2led_an", "p3led_an",
+					"p4led_an", "uart2", "pwm0", "i2s";
+			function = "gpio";
+		};
+	};
+};
+
+&ehci {
+	status = "okay";
+};
+
+&ohci {
+	status = "okay";
+};
+
+&sdhci {
+	status = "okay";
+};
+
+&i2c {
+	status = "okay";
+
+	/* Custom PMIC at 0x0a */
+};
+
+&pcie {
+	status = "okay";
+};
+
+&pcie0 {
+	wifi@0,0 {
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x8000>;
+		ieee80211-freq-limit = <5000000 6000000>;
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <40000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "bootloader";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "config";
+				reg = <0x30000 0x10000>;
+				read-only;
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				label = "loader";
+				reg = <0x50000 0x180000>;
+			};
+
+			partition@1d0000 {
+				label = "params";
+				reg = <0x1d0000 0x10000>;
+				read-only;
+			};
+
+			partition@1e0000 {
+				label = "user_backup";
+				reg = <0x1e0000 0x10000>;
+				read-only;
+			};
+
+			partition@1f0000 {
+				label = "user";
+				reg = <0x1f0000 0x10000>;
+				read-only;
+			};
+
+			partition@200000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x200000 0xdf0000>;
+			};
+
+			partition@ff0000 {
+				label = "mode";
+				reg = <0xff0000 0x10000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&wmac {
+	status = "okay";
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
+};
diff --git a/target/linux/ramips/image/mt76x8.mk b/target/linux/ramips/image/mt76x8.mk
index 7df5c04297693263098549d445ece3fed5d9aba0..97802d2668323ad54d2bb3d7b4f2812911eb6f6e 100644
--- a/target/linux/ramips/image/mt76x8.mk
+++ b/target/linux/ramips/image/mt76x8.mk
@@ -175,6 +175,14 @@ define Device/rakwireless_rak633
 endef
 TARGET_DEVICES += rakwireless_rak633
 
+define Device/ravpower_rp-wd009
+  DTS := RP-WD009
+  IMAGE_SIZE := 14272k
+  DEVICE_TITLE := RAVPower RP-WD009
+  DEVICE_PACKAGES := kmod-mt76x0e kmod-usb2 kmod-usb-ohci kmod-sdhci-mt7620 kmod-i2c-mt7628
+endef
+TARGET_DEVICES += ravpower_rp-wd009
+
 define Device/skylab_skw92a
   DTS := SKW92A
   IMAGE_SIZE := 16064k
